`APIHistoryList`, among other things

CaspianA1 opened the pull request:
***NOTE: NOT READY FOR MERGING YET.***

I was going to contain this to just making an `ApiHistoryList` type, and then implementing spin history with this, and refactoring the Twilio code with this new shared logic. But scope creep invaded, and there's now many more things in this PR. Oh well.

---

### Completed:

(first commit)
- [x] Implement an `ApiHistoryList` type (a cache for image data/general structs associated with entries returned from an API in some historical order; the images/structs are expensive to recompute, so we want to persist them, and not rebuild them for every API call).
- [x] Use the `ApiHistoryList` type to display album history.
- [x] Reduce the number of used Spinitron fields (keeping them commented-out though, since their optionality/existence is important to document - Spinitron does not document what fields are optional, so I had to figure it out on my own).
- [x] Avoid some copying costs in `ContinuallyUpdated`  for Spinitron.
- [x] The streaming server status window, and the clock texture, both load asynchronously with the other grouped-async-loading things in each theme file. This improves load times.
- [x] Lengthen the Spinitron model transition times by a bit.
- [x] Make the default Tokio flavor multithreaded.
- [x] Add various TODOs, improve general readability, and do various refactoring things.

(second commit group)
- [x] If a playlist is an automation playlist, just putting up a note that it's almost done, and not telling the DJ to pack up their things and log out of Spinitron.

(fifth commit)
- [x] Only refreshing Spinitron model's textures now if the 2 model textures' contents would change (before, was refreshing when the model id changed, which resulted in some unnecessary refreshes). A technical byproduct of this was that `TextureCreationInfo` became hashable (which makes implementing [this](https://github.com/WBOR-91-1-FM/wbor-studio-dashboard/issues/3) closer to being possible)).

(eighth commit)
- [x] Add proxying for the weather API (this stops some API overuse). [Here's](https://github.com/WBOR-91-1-FM/wbor-weather-proxy) the code behind that).

---

### Todo:

(first commit; these changes will come after the bigger things)
- [x] Fix the time complexity of the indexed texture retreival from the `ApiHistoryList` type (go from `O(n)` to `O(1)`).
- [x] Migrate the Twilio code to the `ApiHistoryList` type. Perhaps compare Twilio's current output to the `ApiHistoryList` output in some concrete way (how to do?).

(upcoming commit, not pushed yet - this is so much gah - split into bigger and smaller things)

##### Bigger things:
- [x] Retain error reportings in `ContinuallyUpdated`'s `.update` for when a result is still pending.
- [x] Make errors in the error window appear quicker. Can use rapid polling for this, or can perhaps share the error reporting type in different `ContinuallyUpdated` types (via a mutex?). Currently, the speed of error reporting hinges on the shared update rate, rather than the error window's update rate (which is problematic).
- [x] In `ContinuallyUpdated`'s `.update`, only return `true` if new data was recently computed (no pending or error). And for users of `.update`, don't do a view-refresh if it returns `false` (otherwise, can result in bugs; as it had in Spinitron. Check other places too).
- [x] In the Spinitron code, don't initialize the main `SpinitronStateData` before the rendering loop (do it as I go). Use a dummy object as the initial state. Run the continual updater at the same rate, but check for updates more often (to ensure more up-to-data data). This will result in a way faster startup, and in new data appearing more rapidly (and errors possibly being seen quicker once they appear). With this, also check that the spin-history-texture-update only takes 2 update cycles, and not 3.
- [x] Overall, for Spinitron (and for other bits of the code), figure out how the spacing between sync/update will work. Will every sync be matched with a corresponding update? Ensure that that happens; basically, work around this.
- [x] Overall, load less in state-data-initializers corresponding to `ContinuallyUpdated` instances; and just do it asynchronously (e.g. the Twilio phone number, ipinfo location for weather (i.e. the full request URL) (FINISHED WEATHER AND TWILIO BIT), maybe socket loading, and maybe texture data (via a new `LazyTexture` type, or an eager one that starts background loading from the get-go?)).
- [x] Find out when the Spinitron null texture bug happens.

Design goals in the bigger things:
1. Fast startup times
2. Make things pop in as they load (can be further helped through checkpointing for `ContinuallyUpdated`? May implement later). Current plan is to implement this via rapid polling, but could a better way maybe exist? See how callbacks fare with async, and if that will work with the `!Send` constraint of SDL's types. Maybe only rapidly poll when data wasn't gotten when I needed it? I'm not sure. Think about this. 
4. Report errors immediately once they happen
5. Resolve bugs (no-texture bug, and refreshes for when no state changed at all)
6. Making no assumptions about the speed of `ContinuallyUpdated`

##### Smaller things:
- [x] Document all external URLs that are used for requests
- [x] In the end, ensure that double-updates don't happen (happend with weather?) (and think about `i = 0` thing in `update_rate.rs`).
- [x] Load album history concurrently with all other model updates (figure out how to split up that concurrent workload properly)
- [x] Figure out if rate limiting could still arise as a future problem for the weather API (and if showing a message saying "data is stale" will be necessary)
- [x] In the end, ensure that all of this works with the `run_dashboard.sh` script

---

CaspianA1 committed:
This is the first set of changes for the `ApiHistoryList` idea:

spinitron:
- api.rs: shuffling around some functions, and change the wording of `possible` to `maybe`.
- wrapper_types.rs: remove some wrapper types for now.
- model.rs: restructure some of the time parsing code slightly. Also remove a bunch of fields I wasn't using before. Additionally, inline the spin `get` function into a new function that gets the current spin, and the spin history, which saves an API call.
- state.rs: integrate `ApiHistoryList` into the Spinitron state updating (for spins, specifically). Also wrap the cached texture creation info in `Arc`s, to avoid the cost of copying there.

utility_types:
- continually_updated.rs: come up with a really cool idea for intermediate syncing of `ContinuallyUpdated` state.
- mod.rs: add `api_history_list` to this.
- api_history_list.rs: make the first version of this. This will later be improved upon, as I integrate it into the Twilio code. This is perhaps the biggest change here!

dashboard_defs/themes:
- barebones.rs: add a TODO. Add some params for the API history list stuff. Make the streaming server status init run concurrently with the other futures, as well as the clock texture's loading. Lengthen all update intervals to once every 30 seconds.
- standard.rs: ditto (and put the Spinitron windows over the bookshelf).
- retro_room.rs: ditto.

dashboard_defs:
- spinitron.rs: add the API history list windows. Lengthen the transition times by a bit.

main.rs: make the default Tokio flavor multithreaded.

_____

These tasks remain:
- Fixing the time complexity of the API history list index retrieval (it can be made `O(1)`, I think, but it's `O(n)` at the moment).
- Migrating the Twilio code to the API history list idea.

CaspianA1 committed:
Now, for automation playlists, not telling a person to pack up and log out of Spinitron

CaspianA1 committed:
Merge branch 'main' of github.com:WBOR-91-1-FM/wbor-studio-dashboard into CaspianA1/api_history_list

CaspianA1 committed:
Simplify a line

CaspianA1 committed:
Spinitron model image/text texture caching is done properly now.
Each model's 2 associated textures is now only refreshed when the texture and text themselves change, and not the model id (this makes unnecessary transitions not happen anymore).

For various technical reasons, this meant that `TextureCreationInfo` had to become hashable (which was tricky to implement).
This also means that eventually, it will be easier to have text textures as surprises.

Changes:

dashboard_defs:
- error.rs: make a shared fn common.
- spinitron.rs: integrate the new caching behavior.
- all of the themes: use the new hashable hinting newtype.
- twilio.rs: use the new `TextDisplayInfo` wrapper.
- updatable_text_pattern.rs: ditto.
- weather.rs: ditto, and add 2 TODOs.

utility_types:
- mod.rs: add the new hash.rs file.
- hash.rs: make a utility for hashing objects.

texture:
- pool.rs, text.rs: make `TextureCreationInfo` hashable.

spinitron:
- state.rs: implement model texture caching based on when the 2 textures change, rather than only on when the model's ID changes.

CaspianA1 committed:
Make some hashable types more private via better use of the newtype idiom. Make defining hashable newtypes easier too. Also rename `TextMetadata` to `TextMetadataSet`. And, fix the hash function for `HashableTextTextureScrollEaser`.

CaspianA1 committed:
Add a TODO

CaspianA1 committed:
Update the weather code to incorporate a proxy URL that is tried first

CaspianA1 committed:
Change the name RELAY to PROXY. Add a comment about where the proxy code comes from.

CaspianA1 committed:
For api_history_list.rs, add a useful comment and a TODO

CaspianA1 committed:
Make a few lines a bit more readable

mdrxy commented:
What's the Google cal portion?

CaspianA1 commented:
@mdrxy I have this giant list of changes I need to make, scattered through uncommitted code, a local TODO.md document, and various Google calendar reminders (I'll get ideas for features/fixes/etc in the middle of the night, wake up, then make a Google calendar reminder about the idea that appears in the morning). Just trying to unify all of this into one place, to make working on this PR easier.

CaspianA1 committed:
Simplify some logic in continually_updated.rs. Also now only unreporting errors when the error goes away; not when nothing was received (this was happening before).

CaspianA1 committed:
Moved the ipinfo.io API call into the continually updating loop (should improve load times by a bit)

CaspianA1 committed:
Pass some values by value, rather than by reference

CaspianA1 committed:
Make the shared update rate a tad more reasonable again

CaspianA1 committed:
Remove a duplicated struct field

CaspianA1 committed:
Make some async code even more concurrent

CaspianA1 committed:
The fallback texture creation info getter now gets random fallback textures, which should result in more variety

CaspianA1 committed:
Comment out some unused fns for now to avoid spamming the dashboard logs later on

CaspianA1 committed:
Merge branch 'CaspianA1/api_history_list' of github.com:WBOR-91-1-FM/wbor-studio-dashboard into CaspianA1/api_history_list

CaspianA1 committed:
Make an import more idiomatic

CaspianA1 committed:
Get rid of a clone

CaspianA1 committed:
Make some of the surprise code slightly more efficient

CaspianA1 committed:
Get rid of one more clone

CaspianA1 committed:
Move a TODO

CaspianA1 committed:
Changes:

window_tree.rs: add `PartialEq` to `WindowContents`.

dashboard_defs:
- credit.rs: change a param name.
- error.rs: change a param name.
- spinitron.rs: adapt to the new `ContinuallyUpdated` and `SpinitronState` changes.
- streaming_server_status.rs: adapt to the new `ContinuallyUpdated` changes.
- twilio.rs: adapt to the new `ContinuallyUpdated` changes.
- updatable_text_pattern.rs: change a param name.
- weather.rs: adapt to the new `ContinuallyUpdated` changes. Also fix a bug with refreshes happening on the no-weather-info-available string (BUG #1).

spinitron:
- model.rs: make some minor efficiency fixes. Add a TODO. Fix some formatting.
- state.rs: the Spinitron state is now initialized with nothing in it. This speeds up load times by a TON! Also adapt to the new `ContinuallyUpdated` changes; and fix a rare bug with double-refreshes of textures upon network failures (BUG #2).

themes:
- barebones.rs, retro_room.rs, standard.rs: adapt to the new changes.

utility_types:
- api_history_list.rs: avoid a panic (BUG #3).
- continually_updated.rs: see below.

---

There's more details in the code on how the new `ContinuallyUpdated` works, but here are the big ideas:
- The task of "sleeping for N seconds before returning to work" is now delegated to `ContinuallyUpdated`, rather than the users of it. Before, users of it would call `.update` on it once every N seconds, and also refresh the view that often. Now, user of it are free to call it as many times as they want, and update the view much more often. `.update` will not proceed until it's slept for enough time, unless the `.wake_up_if_sleeping` function is called. This means that you don't need to worry about API overuse, and results and errors resulting from it are delivered much, much quicker (calling `.update` may not update its task for another iteration, but if new data is available from its receiving channel, it will update it).
- Basic outcome from this:
1. Less worrying about API call overuse
2. Don't need to worry about calling `.update` every N seconds (the timing is handled internally)
3. Now allowing callers of `.update` to know whether new data is still pending (this depends on the variant of `ContinuallyUpdatedState` which is returned). This allows for figuring out when data is stale, which fixes some bugs with Spinitron (this is BUG #2).
4. Fix a bug in regards to data consistency (if the inner trait function `.update` call fails, then revert the state fully on the task's end - I wasn't doing this properly before) (BUG #4).
5. Make the `new` function not async anymore. This allows me to make a bunch of functions not async anymore throughout the codebase, which is nice.

CaspianA1 committed:
Remove a common thing that pollutes the logs

CaspianA1 committed:
Now loading the Twilio phone number fully in the background, which speeds up dashboard load times even more

CaspianA1 committed:
Spinitron async updating happens even faster now, so it should pop up faster on screen upon boot, and faster otherwise too

CaspianA1 committed:
Spinitron textures are now also loaded concurrently - I think I've reached the maximum amount of optimization I can do here

CaspianA1 committed:
Decrease the view-refresh update rate by more

CaspianA1 committed:
Find a potential bug, and remove the remake transition currently in place for the album history

CaspianA1 committed:
Make Clippy happy

CaspianA1 committed:
Fix some stuff in how the model age data is computed (actually updated now)

CaspianA1 committed:
Change the no-weather-info-available string a bit

CaspianA1 committed:
Load some command results asynchronously for an even faster startup

CaspianA1 committed:
Add a comma I forgot to add

CaspianA1 committed:
Update the Cargo.toml - I forgot to push that before

CaspianA1 committed:
Move a paren down

CaspianA1 committed:
Remove a blank line

CaspianA1 committed:
Remove some Ok wrappers around some constructors

CaspianA1 committed:
Now, after a person's show, a blank persona image is shown (Stewie's outline as the default facebook photo). This resolves the earlier bug when this image was wrongfully carried over.

CaspianA1 committed:
Make a bit of code more readable

CaspianA1 committed:
Change some comment formatting

CaspianA1 committed:
Fix the standard theme color

CaspianA1 committed:
Implement rounded borders! This looks super nice.

CaspianA1 committed:
Make a comment more correct

CaspianA1 committed:
Try to fix the CI

CaspianA1 committed:
Add a bit fat TODO

CaspianA1 committed:
Add a comment

CaspianA1 committed:
Fix a boundary issue in drawing rounded borders

CaspianA1 committed:
Move 2 unwraps into just 1 in a function call

mdrxy committed:
Make Stewie square

CaspianA1 committed:
Add to a comment

CaspianA1 committed:
Merge branch 'CaspianA1/api_history_list' of github.com:WBOR-91-1-FM/wbor-studio-dashboard into CaspianA1/api_history_list

CaspianA1 committed:
Changes:
- Make a util fn in the request code for getting trying multiple fallback URLs for some request
- Use that in the weather code, and with that, add the Spinitron proxy as well (using that mechanism)!
- Fix a bug in `spinitron/state.rs` where the Spinitron show model isn't initialized the first time around
- Fix a major bug in `api_history_list.rs`, where the texture creation info is unnecessarily created many times, leading to possible rate limits (due to fetching content from URLs)

CaspianA1 committed:
Reorder some struct fields

CaspianA1 committed:
Changes:
- Remove the `trigger_surprise.sh` and `trigger_twilio_update.sh` scripts. Replace them with one script that handles all communication to the dashboard, called `communicate_with_dashboard.sh`.
- In this script, allow for instant Spinitron updates as well.
- Greatly simplify the socket reading logic in the Twilio code.

CaspianA1 committed:
Make Stewie wide again - no reason for him to be boxed, really

CaspianA1 committed:
Try to fix some README formatting

CaspianA1 committed:
Try again

CaspianA1 committed:
Last attempt?

CaspianA1 committed:
Remove a break

CaspianA1 committed:
Remove another

CaspianA1 committed:
Remove one last one?

CaspianA1 committed:
Change the error window in the different themes to make the color and position look better

CaspianA1 committed:
Remove a nasty clone by making two sets of fields disjoint for `Window`

CaspianA1 committed:
Make the copy-constraint for ApiHistoryList keys based on clone instead

CaspianA1 committed:
Define the weather api/view refresh update rates outside of the weather code (they're paramaters now)

CaspianA1 committed:
Organize the update rates a bit differently

CaspianA1 committed:
Move 2 vars to the app config file

CaspianA1 committed:
Now adding backticks to the project log for crashes

mdrxy commented:
How close are we to merging?

I think the `Figure out if rate limiting could still arise as a future problem for the weather API (and if showing a message saying "data is stale" will be necessary)` could probably be closed.

CaspianA1 commented:
Roughly 1 week

CaspianA1 committed:
Replace to_generic with to_generic_result

CaspianA1 committed:
Merge branch 'CaspianA1/api_history_list' of github.com:WBOR-91-1-FM/wbor-studio-dashboard into CaspianA1/api_history_list

CaspianA1 committed:
Many changes:

dashboard_defs:
- clock.rs: use `PixelAreaSDL`.
- streaming_server_status.rs: make a reference a value.
- themes/(standard/barebones/retro_room).rs: make the window structure closer to being the same across all the themes. Also change the passed parameters `make_twilio_windows` (this was singular before, too!).
- themes/shared_utils.rs: move a `to_generic_result` out by one layer.
- streaming_server_status.rs: change a `request::get` call to change its new parameter structure.
- weather.rs: minor refactoring.
- twilio.rs: now finally using the `ApiHistoryList` type. This code is so much cleaner now.
- spinitron.rs: slight refactoring.

utility_types:
- generic_result.rs: remove a `pub use`.
- api_history_list.rs: inline `subpool_manager.rs`. Depend more on associated types over generics, to make the code cleaner. Introduce many optimizations to make the code quicker. Also improve the window creation bit of this code.
- time.rs: some general cleanup. Also implement some direct `serde` parsers for rfc2822 and Spinitron timestamps.
- continually_updated.rs: minor tweaks/cleanup.

texture:
- pool.rs: use `PixelAreaSDL`. Make `TextureCreationInfo` not `Clone` anymore (not needed). Use the new `request::get` variant.
- text.rs: use `PixelAreaSDL`. Remove a setter.
- mod.rs: remove `subpool_manager`.
- Also, delete `subpool_manager.rs`. It's now inlined into `api_history_list.rs`.

spinitron:
- model.rs: reorganized model fields a bunch. Use `PixelAreaSDL`. Now auto-deserializing a bunch of start/end times, which means that those start/end times don't have to be reparsed every time.
- api.rs: refactor and clean up. Not using `serde_json::Value` anymore; now directly serializing to the needed model types.
- wrapper_types.rs: move 2 comments.
- state.rs: a bunch of refactors (hard to detail them all).

main.rs: use `PixelAreaSDL`.
window_tree.rs: use `PixelAreaSDL`.
request.rs: avoid rebuilding the client every time; just keep 1 global one (this makes requests WAY faster). Also make `as_type` to a macro, to avoid a fat future monomorphization size.

Cargo.toml: add 2 `serde`-related feature flags to some dependencies.
